src/cheri_insts.sail

```diff
@@ -383,53 +380,37 @@ union clause ast = CAndPerm : (regidx, regidx, regidx)
  * previous value and bits 0 to [cap_hperms_width]-1 of integer register *rs2*
  * and the **uperms** field set to the bitwise and of its previous value and
  * bits [cap_uperms_shift] to [cap_uperms_shift]+[cap_uperms_width]-1 of *rs2*.
- *
- * ## Exceptions
- *
- * An exception is raised if:
- *   - *cs1*.**tag** is not set.
- *   - *cs1* is sealed.
  */
 function clause execute(CAndPerm(cd, cs1, rs2)) = {
   let cs1_val = C(cs1);
   let rs2_val = X(rs2);
-  if not (cs1_val.tag) then {
-    handle_cheri_reg_exception(CapEx_TagViolation, cs1);
-    RETIRE_FAIL
-  } else if isCapSealed(cs1_val) then {
-    handle_cheri_reg_exception(CapEx_SealViolation, cs1);
-    RETIRE_FAIL
-  } else {
-    let perms = getCapPerms(cs1_val);
-    let mask = truncate(rs2_val, cap_perms_width);
-    let newCap = setCapPerms(cs1_val, (perms & mask));
-    C(cd) = newCap;
-    RETIRE_SUCCESS
-  }
+
+  let inCap = clearTagIfSealed(cs1_val);
+  let perms = getCapPerms(inCap);
+  let mask = truncate(rs2_val, cap_perms_width);
```

Comment:
I feel it'd be better to getCapPerms on cs1_val and do the clearTagIfSealed with the setCapPerms (whether inline or just moving the let)?

Reply:
Hmm, I thought of the "clearTagIfSealed" as like a guard that stops any authority leaking into the semantics, which is why initially I put it with the register read, so I was inclined to use `inCap` as much as possible. I guess you want to just put it on the main path through the function, which I can see. I'll give it a go and see how it looks.

Reply:
Yeah, my view is it's logically on the "I'm mutating the capability now" output path, not on the input path